crates/factor-outbound-http/src/wasi.rs

```diff
@@ -104,21 +118,28 @@ impl<'a> WasiHttpView for WasiHttpImplInner<'a> {
 async fn send_request_impl(
     mut request: Request<wasmtime_wasi_http::body::HyperOutgoingBody>,
     mut config: wasmtime_wasi_http::types::OutgoingRequestConfig,
-    allowed_hosts: OutboundAllowedHosts,
+    outbound_allowed_hosts: OutboundAllowedHosts,
+    self_request_origin: Option<SelfRequestOrigin>,
     tls_client_config: Arc<ClientConfig>,
 ) -> anyhow::Result<Result<IncomingResponse, ErrorCode>> {
-    let allowed_hosts = allowed_hosts.resolve().await?;
-
-    let is_relative_url = request.uri().authority().is_none();
-    if is_relative_url {
+    if request.uri().authority().is_some() {
+        // Absolute URI
+        let is_allowed = outbound_allowed_hosts
+            .check_url(&request.uri().to_string(), "https")
+            .await
+            .map_err(|_| ErrorCode::HttpRequestUriInvalid)?;
+        if !is_allowed {
+            return Ok(Err(ErrorCode::HttpRequestDenied));
```

Comment:
~I think we're missing a call to `outbound_allowed_hosts.report_disallowed_host` here, no?~

I see later on that `report_disallowed_host` gets called automatically in `check_url`. Doing it in the `check_url` call but not in `allows_relative_url` is confusing. I think we should encapsulate the call to `report_disallowed_host` in the same equivalent places for both relative and non-relative requests.

Reply:
I reworked this significantly; see the latest commit.